sui/src/rest_gateway.rs

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use move_core_types::identifier::Identifier;
+use move_core_types::language_storage::TypeTag;
+use serde::de::DeserializeOwned;
+use serde_json::{json, Value};
+
+use sui_core::gateway_state::gateway_responses::TransactionResponse;
+use sui_core::gateway_state::GatewayAPI;
+use sui_types::base_types::{encode_bytes_hex, ObjectID, ObjectRef, SuiAddress};
+use sui_types::messages::{Transaction, TransactionData};
+use sui_types::object::ObjectRead;
+
+use crate::rest_gateway::requests::{
+    CallRequest, MergeCoinRequest, PublishRequest, SplitCoinRequest,
+};
+use crate::rest_gateway::responses::{NamedObjectRef, ObjectResponse, TransactionBytes};
+use async_trait::async_trait;
+pub mod requests;
+pub mod responses;
+
+pub struct RestGatewayClient {
+    pub url: String,
+}
+#[async_trait]
+#[allow(unused_variables)]
+impl GatewayAPI for RestGatewayClient {
+    async fn execute_transaction(
+        &mut self,
+        tx: Transaction,
+    ) -> Result<TransactionResponse, anyhow::Error> {
+        let url = format!("{}/api/execute_transaction", self.url);
+        let data = tx.data.to_base64();
+        let sig_and_pub_key = format!("{:?}", tx.tx_signature);
+        let split = sig_and_pub_key.split('@').collect::<Vec<_>>();
+        let signature = split[0];
+        let pub_key = split[1];
+
+        let body = json!({
```

Comment:
Why don't we use strongly typed `SignedTransaction` here and do a `into()`? So that we can always ensure the client implementation is consistent with the API definition

Reply:
we can indeed use SignedTransaction and convert it to json automatically, I am intended to leave one of the method using the raw json construct just for demonstration purpose (more comment and doc WIP), this is mainly for user trying to interface with the rest server using other languages which might not have access to the object model.

Reply:
I will take note of that and provide a Typescript implementation, so probably no need to have it just for demo purpose. It should be straightforward for people to figure out since it's just json object from the OpenAPI spec